flock_agent/daemon/flock_logs.py

```diff
@@ -0,0 +1,140 @@
+import os
+import time
+import json
+
+from .api_client import FlockApiClient
+
+
+class FlockLog:
+    def __init__(self, common, lib_dir):
+        self.c = common
+        self.filename = os.path.join(lib_dir, "flock.log")
+        self.c.log("FlockLog", "__init__", self.filename)
+
+        # If the log file doesn't exist, create an empty one
+        if not os.path.exists(self.filename):
+            open(self.filename, "a").close()
+            os.chmod(self.filename, 0o600)
+
+    def log(self, flock_log_type, twig_ids=None):
+        with open(self.filename, "a") as f:
+            f.write(
+                json.dumps(
+                    {
+                        "type": flock_log_type,
+                        "twig_ids": twig_ids,
+                        "timestamp": int(time.time()),
+                    }
+                )
+                + "\n"
+            )
+        os.chmod(self.filename, 0o600)
+
+    def submit_logs(self):
+        # Keep track of the biggest timestamp we see
+        biggest_timestamp = self.c.global_settings.get("last_flock_log_timestamp")
+
+        # What's the results file's modified timestamp, before we start the import
+        try:
+            mtime = os.path.getmtime(self.filename)
+
+            # Load the log file
+            with open(self.filename, "r") as f:
+                lines = f.readlines()
+                if len(lines) > 0:
+                    self.c.log("FlockLog", "submit_logs", "{} lines".format(len(lines)))
+
+                    # Start an API client
+                    api_client = FlockApiClient(self.c)
+                    try:
+                        api_client.ping()
+                    except:
+                        self.c.log(
+                            "FlockLog",
+                            "submit_logs",
+                            "API is not configured properly",
+                            always=True,
+                        )
+                        return
+
+                    # Make a list of logs
+                    logs = []
+                    for line in lines:
+                        line = line.strip()
+                        try:
+                            obj = json.loads(line)
+                            if "timestamp" in obj:
+                                if "type" not in obj:
+                                    obj["type"] = "unknown"
+
+                                # If we haven't submitted this yet
+                                if obj["timestamp"] > self.c.global_settings.get(
```

Comment:
I don't think file corruption is an issue because of how asyncio works -- it runs everything in a single process, in an event loop, so it's more similar to like async in javascript than to multiple threads. Only one line of code ever runs at a time, so there should never be a situation where the log is getting written to twice simultaneously.

I'm not really concerned with out of order either. Missing a notification is a bigger concern, but I like the idea of increasing the resolution to milliseconds to solve it. I'll do that.